pkg/controllers/nodeclass/garbagecollection/controller.go

```diff
@@ -0,0 +1,155 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package garbagecollection
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/awslabs/operatorpkg/serrors"
+	"github.com/awslabs/operatorpkg/singleton"
+	"k8s.io/apimachinery/pkg/util/sets"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	nodeclaimutils "sigs.k8s.io/karpenter/pkg/utils/nodeclaim"
+
+	v1 "github.com/aws/karpenter-provider-aws/pkg/apis/v1"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/instanceprofile"
+)
+
+type Controller struct {
+	kubeClient              client.Client
+	cloudProvider           cloudprovider.CloudProvider
+	instanceProfileProvider instanceprofile.Provider
+	region                  string
+}
+
+func NewController(kubeClient client.Client, cloudProvider cloudprovider.CloudProvider, instanceProfileProvider instanceprofile.Provider, region string) *Controller {
+	return &Controller{
+		kubeClient:              kubeClient,
+		cloudProvider:           cloudProvider,
+		instanceProfileProvider: instanceProfileProvider,
+		region:                  region,
+	}
+}
+
+func (c *Controller) getActiveProfiles(ctx context.Context) (sets.Set[string], error) {
+	nodeClaims, err := nodeclaimutils.ListManaged(ctx, c.kubeClient, c.cloudProvider)
+	if err != nil {
+		return nil, fmt.Errorf("listing nodeclaims, %w", err)
+	}
+
+	activeProfiles := sets.New[string]()
+	for _, nc := range nodeClaims {
+		if profileName, ok := nc.Annotations[v1.AnnotationInstanceProfile]; ok {
+			activeProfiles.Insert(profileName)
+		}
+
+	}
+
+	return activeProfiles, nil
+}
+
+func (c *Controller) getCurrentProfiles(ctx context.Context) (sets.Set[string], error) {
+	nodeClasses := &v1.EC2NodeClassList{}
+	if err := c.kubeClient.List(ctx, nodeClasses); err != nil {
+		return nil, fmt.Errorf("listing nodeclasses, %w", err)
+	}
+
+	currentProfiles := sets.New[string]()
+
+	for _, nc := range nodeClasses.Items {
+		if nc.Status.InstanceProfile != "" {
+			currentProfiles.Insert(nc.Status.InstanceProfile)
+		}
+	}
+	return currentProfiles, nil
+}
+
+func (c *Controller) shouldDeleteProfile(profileName string, currentProfiles sets.Set[string], activeProfiles sets.Set[string]) bool {
+	// Skip if this is a current profile in any EC2NodeClass
+	if _, ok := currentProfiles[profileName]; ok {
+		return false
+	}
+
+	// Skip if this is a protected profile
+	if c.instanceProfileProvider.IsProtected(profileName) {
+		return false
+	}
+
+	// Skip if this is an active profile (NodeClaims are using it)
+	if _, isActive := activeProfiles[profileName]; isActive {
```

Comment:
same suggestion here to use `Has`
```suggestion
	if activeProfiles.Has(profileName) {
```